### PR TITLE
src/msg/async/rdma: decrease cpu usage by rdtsc instruction

### PR DESCRIPTION
**Pull request by TsaiJin**
The `RDMADispatcher::polling()` thread consumes CPU a lot. 
We can get the cpu time distribution stack of the `RDMADispatcher::polling()` thread with the help of awesome gdbprof tool:
```
+ 100.00% ???
      + 100.00% RDMADispatcher::polling
        + 42.40% Infiniband::CompletionQueue::poll_cq
        | + 37.20% ibv_poll_cq
        |   + 37.20% mlx5_poll_cq_1
        |     + 18.00% pthread_spin_lock
        |     + 1.20% pthread_spin_lock@plt
        |     + 0.80% pthread_spin_unlock
        |     + 0.40% mlx5_free_srq_wqe
        |       + 0.40% pthread_spin_lock
        + 28.40% ceph_clock_now
        | + 26.00% clock_gettime
        | | + 23.60% __vdso_clock_gettime
        | |   + 20.80% do_realtime
        | |   | + 17.20% vgetsns
        | |   | | + 12.00% vread_tsc
        | |   | | | + 11.20% vget_cycles
        | |   | | |   + 11.20% __native_read_tsc
        | |   | | + 4.40% vread_pvclock
        | |   | + 2.40% read_seqcount_begin
        | |   | | + 2.40% __read_seqcount_begin
        | |   | + 0.80% timespec_add_ns
        | |   + 0.40% do_monotonic
        | |     + 0.40% timespec_add_ns
        | + 0.40% clock_gettime@plt
        + 16.80% PerfCounters::set
        + 2.80% RDMAConnectedSocketImpl::pass_wc(std::vector<ibv_wc, std::allocator<ibv_wc> >&&)
        | + 2.40% RDMAConnectedSocketImpl::notify
        | | + 2.40% write
        | + 0.40% Locker
        |   + 0.40% Mutex::Lock
        |     + 0.40% pthread_mutex_lock
        |       + 0.40% _L_lock_791
        |         + 0.40% __lll_lock_wait
        + 1.20% poll
        | + 1.20% poll
        + 1.20% operator-
        | + 1.20% utime_t
        |   + 1.20% normalize
        + 0.80% to_nsec
        + 0.80% operator std::__atomic_base<unsigned long>::__int_type
        | + 0.80% load
        + 0.40% PerfCounters::inc
```

From the stack we can figure out that the function `ceph_clock_now()` takes about **30%** CPU time. 

The function `ceph_clock_now()` has a nanosecond resolution(a call of `clock_gettime(CLOCK_REALTIME)`) and each execution takes 73 CPU cycles (from a core of *Intel(R) Xeon(R) CPU E5-2650 v2 @ 2.60GHz*).
 However, we just need microsecond resolution and time increment instead of wall clock time in the polling thread. So use `rdtsc` will decrease the CPU usage sharply.

The CPU time distribution stack by using `rdtsc` is as following. Compared with `ceph_clock_now()`, the function `Cycles::to_microseconds` takes about **22%** CPU time.
```
+ 100.00% ???
      + 100.00% RDMADispatcher::polling
        + 44.00% Infiniband::CompletionQueue::poll_cq
        | + 41.00% ibv_poll_cq
        |   + 40.50% mlx5_poll_cq_1
        |     + 17.00% pthread_spin_lock
        |     + 1.00% pthread_spin_unlock
        |     + 0.50% pthread_spin_unlock@plt
        + 22.00% PerfCounters::set
        | + 0.50% operator=
        |   + 0.50% store
        + 16.00% Cycles::to_microseconds
        | + 12.00% Cycles::to_nanoseconds
        + 6.00% rdtsc
        + 3.00% RDMAConnectedSocketImpl::pass_wc(std::vector<ibv_wc, std::allocator<ibv_wc> >&&)
        | + 2.00% Locker
        | | + 2.00% Mutex::Lock
        | |   + 2.00% pthread_mutex_lock
        | |     + 2.00% _L_lock_791
        | |       + 2.00% __lll_lock_wait
        | + 0.50% insert<__gnu_cxx::__normal_iterator<ibv_wc*, std::vector<ibv_wc> >, void>
        | | + 0.50% _M_insert_dispatch<__gnu_cxx::__normal_iterator<ibv_wc*, std::vector<ibv_wc> > >
        | |   + 0.50% std::vector<ibv_wc, std::allocator<ibv_wc> >::_M_range_insert<__gnu_cxx::__normal_iterator<ibv_wc*, std::vector<ibv_wc, std::allocator<ibv_wc> > > >
        | |     + 0.50% _M_deallocate
        | |       + 0.50% deallocate
        | |         + 0.50% free
        | |           + 0.50% ifree
        | |             + 0.50% je_isalloc
        | |               + 0.50% je_arena_salloc
        | |                 + 0.50% je_arena_mapbits_binind_get
        | + 0.50% RDMAConnectedSocketImpl::notify
        |   + 0.50% write
        + 2.00% poll
        | + 1.50% poll
        + 1.50% Infiniband::QueuePair::~QueuePair
        | + 1.50% mlx5_destroy_qp
        |   + 1.00% ibv_cmd_destroy_qp
        |   | + 1.00% write
        |   + 0.50% mlx5_free_actual_buf
        |     + 0.50% munmap
        + 0.50% RDMADispatcher::handle_tx_event
        | + 0.50% RDMADispatcher::post_tx_buffer
        |   + 0.50% RDMADispatcher::notify_pending_workers
        |     + 0.50% pop_front
        |       + 0.50% _M_erase
        |         + 0.50% _M_put_node
        |           + 0.50% deallocate
        |             + 0.50% free
        |               + 0.50% ifree
        |                 + 0.50% je_iqalloc
        |                   + 0.50% je_iqalloct
        |                     + 0.50% je_idalloct
        |                       + 0.50% je_arena_dalloc
        |                         + 0.50% je_tcache_dalloc_small
        |                           + 0.50% je_tcache_bin_flush_small
        |                             + 0.50% je_malloc_mutex_lock
        |                               + 0.50% pthread_mutex_lock
        |                                 + 0.50% _L_lock_791
        |                                   + 0.50% __lll_lock_wait
        + 0.50% PerfCounters::inc
```

Signed-off-by: Jin Cai <caijin.caij@alibaba-inc.com>

### CODE REVIEW

**Comment by TsaiJin** [comment]
@liewegas Expect to get your advice.

File: src/msg/async/rdma/RDMAStack.cc [hunk 1]
```diff
@@ -245,7 +247,8 @@ void RDMADispatcher::polling()
       if (!num_qp_conn && done)
         break;
 
-      if ((ceph_clock_now() - last_inactive).to_nsec() / 1000 > cct->_conf->ms_async_rdma_polling_us) {
+      uint64_t now = Cycles::rdtsc();
+      if (Cycles::to_microseconds(now - last_inactive) > cct->_conf->ms_async_rdma_polling_us) {
```
Comment (branch-predictor) [inline src/msg/async/rdma/RDMAStack.cc]: Use Cycles::to_nanoseconds() and use pre-calculated value of cct->_conf->ms_async_rdma_polling_us * 1000 to get rid of one floating point division, or better yet, precompute value of cct->_conf->ms_async_rdma_polling_us in cycles and compare without using Cycles::to_microseconds at all.
In any case, using RDTSC is prone to errors resulting from CPU frequency scaling and differing TSC values when thread switches between CPU cores. Not sure if reducing CPU usage (from 28% to 22%) is really worth the risk.

Comment (TsaiJin) [inline src/msg/async/rdma/RDMAStack.cc]: @branch-predictor Yes, the problems of RDTSC mentioned by you above are also my concerns.  
 Even though the difference of TSC between CPUs can be solved by pin the polling thread to a specified core, I don't think it is elegant. 
Do you have any other suggestions to reduce the CPU usage of RDMA polling thread?

Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAStack.cc]: I think it's ok. potential microsecond inaccurate is fine in this case

Comment (TsaiJin) [inline src/msg/async/rdma/RDMAStack.cc]: @yuyuyu101 Yes, we only need microsecond resolution.

**Pull request merged**